### PR TITLE
GLT-2189: allow admins to manage Tissue Origins

### PR DESCRIPTION
**Pull request by djcooke**
Bonus: server-side validation now results in nicer messages in handsontable pages

### CODE REVIEW

**Comment by sqwci** [comment]
SonarQube analysis reported 4 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TissueOriginController.java#L71](https://github.com/TGAC/miso-lims/blob/70b11749315a3f9c018ea1696440215c1136e127/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java#L71): Remove this useless assignment; "dto" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MAJOR][MAJOR] [TissueOriginController.java#L93](https://github.com/TGAC/miso-lims/blob/70b11749315a3f9c018ea1696440215c1136e127/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java#L93): Remove this useless assignment; "tissueOriginDto" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/70b11749315a3f9c018ea1696440215c1136e127/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/70b11749315a3f9c018ea1696440215c1136e127/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultTissueOriginService.java [hunk 1]
```diff
@@ -38,25 +45,67 @@ public TissueOrigin get(Long tissueOriginId) throws IOException {
...
+   * @throws IOException
+   */
+  private void setChangeDetails(TissueOrigin tissueOrigin) throws IOException {
     User user = authorizationManager.getCurrentUser();
-    updatedTissueOrigin.setUpdatedBy(user);
-    tissueOriginDao.update(updatedTissueOrigin);
+    Date now = new Date();
+    tissueOrigin.setUpdatedBy(user);
+
+    if (tissueOrigin.getId() == Array.UNSAVED_ID) {
```
Comment (apmasell) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultTissueOriginService.java]: Why is this `Array`?

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java [hunk 2]
```diff
@@ -98,32 +96,27 @@ private static TissueOriginDto writeUrls(TissueOriginDto tissueOriginDto, UriCom
...
-    return new ResponseEntity<>(HttpStatus.OK);
+    return Dtos.asDto(tissueOriginService.get(id));
   }
 
   @RequestMapping(value = "/tissueorigin/{id}", method = RequestMethod.DELETE)
-  @ResponseBody
-  public ResponseEntity<?> deleteTissueOrigin(@PathVariable("id") Long id, HttpServletResponse response) throws IOException {
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deleteTissueOrigin(@PathVariable("id") Long id, HttpServletResponse response) throws IOException {
     tissueOriginService.delete(id);
```
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java]: Could it return a friendlier error message if the tissue origin is in use? Current message is 
```
could not execute statement; SQL [n/a]; constraint [null]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
```

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 3]
```diff
@@ -577,97 +577,100 @@ HotTarget.sample = (function() {
...
-              alert("You have selected samples of classes " + classes.map(Utils.array.getAlias).join(" & ")
-                  + ". Please select samples from only one tissue processing class.");
-              return;
-            }
+              var classes = getSampleClasses(samples);
+              var categories = Utils.array.deduplicateString(classes.map(function(sampleClass) {
+                return sampleClass.sampleCategory;
+              }));
+              if (categories.length > 1) {
+                alert("You have selected samples of categories " + categories.join(" & ")
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 4]
```diff
@@ -577,97 +577,100 @@ HotTarget.sample = (function() {
...
-              targets.push({
-                name: "Library",
-                action: function(replicates) {
-                  window.location = "/miso/library/bulk/propagate?" + jQuery.param({
-                    ids: idsString,
-                    replicates: replicates
-                  });
-                }
+              if (categories[0] == 'Tissue Processing' && classes.length > 1) {
+                alert("You have selected samples of classes " + classes.map(Utils.array.getAlias).join(" & ")
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/TissueOrigin.java [hunk 5]
```diff
@@ -5,12 +5,16 @@
 
 import com.eaglegenomics.simlims.core.User;
 
-public interface TissueOrigin extends Serializable {
+public interface TissueOrigin extends Aliasable, Serializable {
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/TissueOrigin.java]: CRITICAL Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/AbstractInstituteDefaultsController.java [hunk 6]
```diff
@@ -113,6 +115,8 @@ public void setSecurityManager(SecurityManager securityManager) {
   }
 
   protected void writeConfiguration(ObjectMapper mapper, ObjectNode config) throws IOException {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/AbstractInstituteDefaultsController.java]: MAJOR Remove this unused method parameter "mapper". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

**Comment by sqwci** [comment]
SonarQube analysis reported 20 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 8 major
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L219](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L219): Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L227](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L227): Define a constant instead of duplicating this literal "Detailed sample is missing parent identifier" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L86](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L86): A "NullPointerException" could be thrown; "parent" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L499](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L499): String contains no format specifiers. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [AbstractInstituteDefaultsController.java#L36](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/AbstractInstituteDefaultsController.java#L36): Remove this unused "log" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [TissueOriginController.java#L71](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java#L71): Remove this useless assignment; "dto" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MAJOR][MAJOR] [TissueOriginController.java#L93](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/TissueOriginController.java#L93): Remove this useless assignment; "tissueOriginDto" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [DefaultSampleService.java#L408](https://github.com/TGAC/miso-lims/blob/5c0a1e3222c8a3e2e4750cc56b9e1308b445364a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L408): Remove the declaration of thrown exception 'org.hibernate.exception.ConstraintViolationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**